### ISSUE TITLE
code

### ISSUE THREAD

**Issue by yuMMM125** [higokri/Reimplementation-of-PromptSG#1]
hello, have you reconstructed the code of promptSG?

**Comment by higokri** [comment]
I am a student trying to reproduce the Promptsg paper. Since my code does not achieve the performance reported in the paper, I have created a GitHub repository to discuss and identify any mistakes. I will upload my code as soon as it is organized. Thank you! :)

**Comment by yuMMM125** [comment]
I'm also trying to reconstruct the code. But the results were not good. For example, once I add the inversion network to the model, the mAP will decrease. And since the output feature length of cross-attention is 1, I'm curious about what the input of self-attention should be. Have you ever encountered these problems?  And how did you solve them? Thank you！

**Comment by higokri** [comment]
Yes, in the paper, it seems that they used EOT(feature) [B,1,512] as the query when inputting text features into the Transformer Multi-Fusion Block. I found this a bit odd, so I tried using the entire [B,L,512] as the query instead. Then, for Re-ID, I utilized the feature at the EOT index from the output of the Multi-Fusion Block. However, this approach resulted in a performance drop. So, I’m considering reaching out to the authors via email to clarify this.

**Comment by flin2002** [comment]
I'm also working on this. 89% mAP on Market1501 is the highest performance of my reimplementation. Seeking for help. ;(

**Comment by 88Peggy** [comment]
你好，请问哪里有这篇论文的代码呀555

**Comment by ZYK23** [comment]
我也是遇见了这个问题，我尝试了复现，使用CLIP-REID作为基线，并去掉state1阶段训练。三层linear组成反转网络，使用文本eos token或整个77长度token作为查询，使用图像class token或全部token作为键个值。输入到复现的多头（8头）交叉注意力和两层自注意力，发现效果均没有PromptSG效果好（差距很大），此外，也没有CLIP-ReID效果好。

**Comment by ZYK23** [comment]
> I'm also working on this. 89% mAP on Market1501 is the highest performance of my reimplementation. Seeking for help. ;(

The result of my reproduction is only 83map, which is based on CLIP-ReID. Excuse me, what are your details?

**Comment by ZYK23** [comment]
> I'm also working on this. 89% mAP on Market1501 is the highest performance of my reimplementation. Seeking for help. ;(

self.inversion_network = nn.Sequential(
            nn.Linear(768, ctx_dim), # CLIP图像编码器最终需要768投影到512，这里取投影前的维度，在投影前传入到反转网络
            nn.ReLU(),
            nn.Linear(ctx_dim, ctx_dim),
            nn.ReLU(),
            nn.Linear(ctx_dim, ctx_dim),
            nn.BatchNorm1d(ctx_dim)；

class ResidualAttentionBlock(nn.Module):
    def __init__(self, d_model: int, n_head: int):
        super().__init__()

        self.attn = nn.MultiheadAttention(d_model, n_head, dropout=0.1)
        self.ffn = nn.Sequential(
                            nn.Linear(d_model, d_model * 4),
                            nn.ReLU(),
                            nn.Linear(d_model * 4, d_model),
                            nn.Dropout(0.1))
        
        self.norm1 = nn.LayerNorm(d_model, eps=1e-12)
        self.norm2 = nn.LayerNorm(d_model, eps=1e-12)

    def attention(self, x: torch.Tensor):
        return self.attn(x, x, x)[0]

    def forward(self, x: torch.Tensor):
        attn_out = self.attention(x)
        x = x + attn_out              # 注意力残差
        x = self.norm1(x)             # LN

        ffn_out = self.ffn(x)
        x = x + ffn_out               # FFN 残差
        x = self.norm2(x)             # LN
        return x
    
class MultimodalInteractionModule(nn.Module):
    def __init__(self, embed_dim=512, num_heads=8, num_transformer_blocks=2):
        super().__init__()
        self.embed_dim = embed_dim
        
        self.q_proj = nn.Linear(embed_dim, embed_dim)  # 文本→Q
        self.k_proj = nn.Linear(embed_dim, embed_dim)  # patch→K
        self.v_proj = nn.Linear(embed_dim, embed_dim)  # patch→V
        self.cross_attn = nn.MultiheadAttention(embed_dim, num_heads, dropout=0.1)
        self.transformer_blocks = nn.Sequential(*[ResidualAttentionBlock(embed_dim, num_heads) for _ in range(num_transformer_blocks)])


    def forward(self, patch_emb, text_emb, tokenized_prompts):
        
        # patch_emb=[b,128,512]
        # text_emb = [b,77,512]
        B,L,D = patch_emb.shape
        text_emb = text_emb[torch.arange(text_emb.shape[0]), tokenized_prompts.argmax(dim=-1)]

        q = self.q_proj(text_emb.unsqueeze(1))
        k = self.k_proj(patch_emb[:,1:,:])
        v = self.v_proj(patch_emb[:,1:,:])

        q = q.permute(1, 0, 2)  # NLD -> LND  
        k = k.permute(1, 0, 2)  # NLD -> LND  
        v = v.permute(1, 0, 2)  # NLD -> LND 

        x = self.cross_attn(q,k,v)[0] # [b,1,512]
        x = self.transformer_blocks(x)
        x = x.permute(1, 0, 2)  # LND -> NLD 

        return x.squeeze(1)；

It seems that the important things are the text inversion network and the multimodal interaction module.
